apps/aecore/src/aec_chain.erl

```diff
@@ -723,76 +723,67 @@ get_key_block_by_height(Height) when is_integer(Height), Height >= 0 ->
 
 -spec get_current_generation() -> 'error' | {'ok', generation()}.
 get_current_generation() ->
-    get_generation_(top_block_node()).
-
-get_generation_(#{hash := Hash, header := Header}) ->
-    case aec_headers:type(Header) of
-        key ->
-            {ok, #{ key_block => aec_blocks:new_key_from_header(Header), micro_blocks => [], dir => forward }};
-        micro ->
-            Index = maps:from_list(aec_db:find_headers_and_hash_at_height(aec_headers:height(Header))),
-            get_generation_with_header_index(Hash, Index)
-    end;
-get_generation_(Hash) when is_binary(Hash) ->
-    case get_header(Hash) of
-        error -> error;
-        {ok, Header} ->
-            get_generation_(#{hash => Hash, header => Header})
-    end;
-get_generation_(_) ->
-    error.
-
-get_generation_with_header_index(TopHash, Index) ->
-    get_generation_with_header_index(TopHash, Index, []).
-
-get_generation_with_header_index(TopHash, Index, MBs) ->
-    H = maps:get(TopHash, Index),
-    case aec_headers:type(H) of
-        key -> {ok, #{ key_block => aec_blocks:new_key_from_header(H), micro_blocks => MBs, dir => forward }};
-        micro ->
-            Block = aec_db:get_block_from_micro_header(TopHash, H),
-            get_generation_with_header_index(aec_headers:prev_hash(H), Index, [Block | MBs])
-    end.
+    #{header := TopHeader} = top_block_node(),
+    get_generation_by_height(aec_headers:height(TopHeader), forward).
 
 -spec get_generation_by_hash(binary(), forward | backward) ->
         'error' | {'ok', generation()}.
 get_generation_by_hash(KeyBlockHash, Dir) ->
     case aec_db:find_key_block(KeyBlockHash) of
         none -> error;
-        {value, KeyBlock} when Dir == backward ->
-            case get_generation_(aec_blocks:prev_hash(KeyBlock)) of
-                error         -> error;
-                {ok, G = #{}} -> {ok, G#{ key_block => KeyBlock, dir => Dir }}
-            end;
-        {value, KeyBlock} when Dir == forward ->
-            case get_generation_by_height(aec_blocks:height(KeyBlock), forward) of
-                {ok, G = #{ key_block := KeyBlock }} -> {ok, G};
```

Comment:
I have a hard time following where this logic has gone and why it was here at the first place.

I assume that somehow this would kick in if you are on a fork, get a hash of a key block, then start searching, but the generation you get returned does no longer contain the keyblock you started of from, due to the fork??

Reply:
This logic has not gone anywhere - I think it is an artefact of the convoluted way this function was implemented before. I.e. to get the generation _forward_ from a key-block(hash) you basically grabbed the _next_ key-block and then fetched a generation backwards from that. 

In effect that makes this a check that the keyblock-hash is on the current chain. Not sure if that is intended or not, but it is a good observation, I think we can keep the old behavior - I'll add an extra check.